### PR TITLE
Add member auth to the Delivery API

### PR DESCRIPTION
**Pull request by kjac**
### Prerequisites

- [x] I have added steps to test this contribution in the description below

### Description

This PR adds member auth to the Delivery API. The PR is targeted v14 on purpose, because we need to ensure that the solution can co-exist with users auth in the Management API. A separate task exists to copy all relevant parts of this PR to v12 once this PR is merged.

This "backwards" kind of developing means I've tried to target a v12 PR in a v14 PR. It gets a little weird in some places. For example: 

- Some obsoletion messages in this PR target v14, because they're going to be copied over to the v12 branch.
- Some changes appear to be terribly breaking, but they affect only code that is currently part of v14 (like the removed `Paths` and the changed `OAuthClientIds`), so they're OK to merge in.

At this time, member auth is not expected to be widely adopted by consumers of the Delivery API. Therefore this PR has been constructed to have zero negative impact at query time, if member auth is not used in the Delivery API. This results in a few somewhat wonky code execution paths, but as the alternative would be to inflict a performance penalty on every last consumer of the Delivery API, the wonky paths are a fair trade-off.

### Testing this PR

This PR should be tested alongside the docs article that is being finalized now. The article contains a lot of code samples that are required to test the PR.

It's not going to make a lot of sense to outline every last test scenario here, as there are a LOT and most all of them require additional test projects on the side to perform auth against the Delivery API. Thus, testing will be performed as peer testing with @kjac 

A little checklist of things to absolutely remember (mostly for my own sake):

- [x] custom cookie options login path
- [x] member auth can be enabled and disabled by config
- [x] protected content is not indexed if member auth is disabled
- [x] external auth is possible (i.e. GitHub)
- [x] hybrid/combined auth also works
- [x] multiple return URLs can be used, also with external auth
- [x] user auth cannot authenticate members and vice versa
- [x] member tokens are revoked when member role membership changes
- [x] locked out, unapproved and deleted members cannot use previously issued tokens
- [x] refresh tokens work
- [x] can perform token revoke
- [x] can perform sign-out from the server
- [x] can query protected content when authorized
- [x] cannot query protected content as anonymous
- [x] can fetch protected content (single items) when authorized
- [x] cannot fetch protected content (single items) as anonymous
- [x] must use redirect URLs that are on the allow-list when logging in
- [x] must use redirect URLs that are on the allow-list when logging out
- [x] can enable Swagger auth for members

### CODE REVIEW

File: src/Umbraco.Cms.Api.Delivery/Controllers/Security/MemberController.cs [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+        }
+
+        return request.IdentityProvider.IsNullOrWhiteSpace()
+            ? await AuthorizeInternal(request)
+            : await AuthorizeExternal(request);
+    }
+
+    [HttpGet("signout")]
+    [MapToApiVersion("1.0")]
+    public async Task<IActionResult> Signout()
```
Comment (elit0451) [inline src/Umbraco.Cms.Api.Delivery/Controllers/Security/MemberController.cs:81]: Same here - no `ProducesResponseType`

**Review by elit0451 — COMMENTED** [review]
Looks great 💪 🙌 , I've left a few comments for you to consider 🙂

File: src/Umbraco.Core/Models/DeliveryApi/ProtectedAccess.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+    public ProtectedAccess(Guid? memberKey, string[]? memberRoles)
+    {
+        MemberKey = memberKey;
+        MemberRoles = memberRoles;
+    }
+
+    public Guid? MemberKey { get; }
+
+    public string[]? MemberRoles { get; }
```
[the remark this answers is not in the document]
Comment (kjac) [reply src/Umbraco.Core/Models/DeliveryApi/ProtectedAccess.cs:15]: I think it's OK to have it explicitly nullable. Might come in handy as a tri-state down the road.

File: src/Umbraco.Cms.Api.Delivery/Controllers/Security/MemberController.cs [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+        _httpContextAccessor = httpContextAccessor;
+        _memberSignInManager = memberSignInManager;
+        _memberManager = memberManager;
+        _logger = logger;
+        _deliveryApiSettings = deliveryApiSettings.Value;
+    }
+
+    [HttpGet("authorize")]
+    [MapToApiVersion("1.0")]
+    public async Task<IActionResult> Authorize()
```
[the remark this answers is not in the document]
Comment (kjac) [reply src/Umbraco.Cms.Api.Delivery/Controllers/Security/MemberController.cs:51]: Nope - same as backoffice. If anything we should hide this controller from Swagger. Noted as a future potential 😄

File: src/Umbraco.Cms.Api.Delivery/Controllers/Security/MemberController.cs (hunk 1 above)
Comment (kjac) [reply src/Umbraco.Cms.Api.Delivery/Controllers/Security/MemberController.cs:81]: Nope - same as backoffice. If anything we should hide this controller from Swagger. Noted as a future potential 😄

File: src/Umbraco.Cms.Api.Delivery/Controllers/Security/MemberController.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (kjac) [reply src/Umbraco.Cms.Api.Delivery/Controllers/Security/MemberController.cs:51]: Update - we'll hide it right away 😆

**Review by elit0451 — APPROVED** [review]
Looks awesome! 🚀🔥

**Review by elit0451 — APPROVED** [review]

**Pull request merged**